Classes/MapEditor/MapEditor.cs

```diff
@@ -391,56 +398,47 @@ public class MapEditor {
+                double offset = 0.0;
+                if (lastBeatChange.globalBeat > 0.0) {
+                    double differenceDefaultNew = Math.Floor(lastBeatChange.globalBeat / defaultGridLength) * defaultGridLength;
+                    offset = lastBeatChange.globalBeat - differenceDefaultNew;
+                }
+
+                double newBeat = Math.Round(n.beat / defaultGridLength) * defaultGridLength + offset;
+                if (Helper.DoubleApproxEqual(n.beat, newBeat) || Helper.DoubleApproxEqual(n.beat, newBeat - defaultGridLength)) {
+                    newBeat = n.beat;
+                }
 
-        foreach (Note n in currentMapDifficulty.selectedNotes) {
+                return new Note(newBeat, n.col);
+            })
+            .ToList();
+        UpdateNotes(quantizedNotes, currentMapDifficulty.selectedNotes);   
+    }
 
-            // find next beat change 
+    public double GetGridLength(double bpm, int gridDivision) {
+        double scaleFactor = globalBPM / bpm;
+        return scaleFactor / gridDivision;
+    }
+
+    public BPMChange GetLastBeatChange(double beat) {
+        if (currentMapDifficulty != null) {
             BPMChange? currentBeat = currentMapDifficulty
                 .bpmChanges
                 .OrderByDescending(obj => obj.globalBeat)
-                .Where(obj => obj.globalBeat <= n.beat)
+                .Where(obj => obj.globalBeat <= beat)
                 .Select(obj => obj)
                 .FirstOrDefault();
-            
-            double offset = 0.0;
-
-            // beat has been changed
-            if (currentBeat != null){
-
-                // calculate the new length factor
-                double scaleFactor = globalBPM / currentBeat.BPM;
-
-                // calculate the length of a single beat based on the new length
-                defaultBeats = scaleFactor / currentBeat.gridDivision;
-
-                // calculate time difference between old beat and start float
-                double differenceDefaultNew = Math.Floor(currentBeat.globalBeat / defaultBeats) * defaultBeats;
-
-                // calculate offset
-                offset = currentBeat.globalBeat - differenceDefaultNew;
-
+            if (currentBeat != null) {
+                return currentBeat;
             }
-            double newBeat = Math.Round(n.beat / defaultBeats) * defaultBeats;
-
-            newBeat += offset;
-
-            // no changes done? or its a step jump? do nothing.
-            if (Helper.DoubleApproxEqual(n.beat, newBeat) || Helper.DoubleApproxEqual(n.beat, newBeat - defaultBeats) ) {
-                newBeat = n.beat;
-            }
-
-            // save changes
-            Note newNote = new Note(newBeat, n.col);
-            notesToAdd.Add(newNote);
-            notesToRemove.Add(n);
         }
-        UpdateNotes(notesToAdd, notesToRemove);   
+        return new BPMChange(0.0, globalBPM, defaultGridDivision);
```

Comment:
You can pass this as a parameter to FirstOrDefault to simplify the logic.

Reply:
Definitely makes sense